### PR TITLE
MGMT-7529 Migrate local auth to use InfraEnv

### PR DESCRIPTION
**Pull request by rollandf**
# Assisted Pull Request

## Description
Migrate local auth to use InfraEnv:
Add the possibility to create JWT token for local auth with claim id with infraenv or cluster id

The option for cluster id is needed for kube API AgentClusterInstall event and logs URL, as AgentClusterInstall does not have a direct link to InfraEnv

Signed-off-by: Fred Rolland <frolland@redhat.com>

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @yevgeny-shnaidman 
/cc @filanov 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by filanov** [comment]
cc  @carbonin

**Comment by filanov** [comment]
Can you explain why we are moving everything to infra env? 
We said that cluster can be created from hosts from multiple Infra envs? so with one you are using? 
In addition events in this case are related to a cluster, so why can't we continue using the cluster id to sign the URL?

**Comment by rollandf** [comment]
> Can you explain why we are moving everything to infra env?
> We said that cluster can be created from hosts from multiple Infra envs? so with one you are using?
> In addition events in this case are related to a cluster, so why can't we continue using the cluster id to sign the URL?

The local auth is used mainly to add an auth layer on the communication between the host and the assisted service.
We also used it to be able to access events and logs.

Since ClusterDeployment does not have a single relation to infraEnv, seems that we need to support both.

File: pkg/auth/local_authenticator.go [hunk 1]
```diff
@@ -61,25 +62,42 @@ func (a *LocalAuthenticator) AuthAgentAuth(token string) (interface{}, error) {
...
 
-	_, exists := a.cache.Get(clusterID)
-	if !exists {
-		if clusterExists(a.db, clusterID) {
-			a.cache.Set(clusterID, "", cache.DefaultExpiration)
-		} else {
-			err := errors.Errorf("cluster %s does not exist", clusterID)
-			a.log.Error(err)
-			return nil, common.NewInfraError(401, err)
+	if infraEnvOk {
```
Comment (filanov) [inline pkg/auth/local_authenticator.go:73]: Can it be that we will check both?

**Comment by filanov** [comment]
@rollandf you need to change the commit message to have `MGMT-7529:`

File: internal/gencrypto/token.go [hunk 2]
```diff
@@ -8,22 +8,29 @@ import (
 	"github.com/pkg/errors"
 )
 
-func LocalJWT(cluster_id string) (string, error) {
+type LocalJWTKeyType string
+
+const (
+	InfraEnvKey LocalJWTKeyType = "infraEnv_id"
```
[the remark this answers is not in the document]
Comment (rollandf) [reply internal/gencrypto/token.go]: done

File: pkg/auth/local_authenticator.go [hunk 3]
```diff
@@ -113,3 +131,8 @@ func clusterExists(db *gorm.DB, clusterID string) bool {
 	var c common.Cluster
 	return db.Select("id").Take(&c, map[string]interface{}{"id": clusterID}).Error == nil
 }
+
+func infraEnvExists(db *gorm.DB, infraEnvID string) bool {
+	var c common.InfraEnv
+	return db.Select("id").Take(&c, map[string]interface{}{"id": infraEnvID}).Error == nil
```
[the remark this answers is not in the document]
Comment (rollandf) [reply pkg/auth/local_authenticator.go]: done

File: pkg/auth/local_authenticator.go (hunk 1 above)
Comment (rollandf) [reply pkg/auth/local_authenticator.go:74]: added `else`

**Review by filanov — APPROVED** [review]

File: internal/controller/controllers/agent_controller.go [hunk 4]
```diff
@@ -304,12 +304,12 @@ func (r *AgentReconciler) populateEventsURL(log logrus.FieldLogger, agent *aiv1b
...
 }
 
-func (r *AgentReconciler) eventsURL(log logrus.FieldLogger, clusterId, agentId string) (string, error) {
+func (r *AgentReconciler) eventsURL(log logrus.FieldLogger, clusterId, infraEnvId, agentId string) (string, error) {
 	eventsURL := fmt.Sprintf("%s%s/v1/clusters/%s/events?host_id=%s", r.ServiceBaseURL, restclient.DefaultBasePath, clusterId, agentId)
 	if r.AuthType != auth.TypeLocal {
 		return eventsURL, nil
 	}
-	eventsURL, err := gencrypto.SignURL(eventsURL, clusterId)
+	eventsURL, err := gencrypto.SignURL(eventsURL, infraEnvId, gencrypto.InfraEnvKey)
```
Comment (carbonin) [inline internal/controller/controllers/agent_controller.go:312]: Why did this events url change to infra env if the other is still using cluster?

Comment (filanov) [reply internal/controller/controllers/agent_controller.go:312]: All the operations related to a host need to have infra env.
With v2 and late binding we will probably need to change this URL based on the current status, for example if host is unbound then we will need to have infra env and host id, when it's bound we will search by host id and cluster. 
So there will be work here once we move to `/v2/events` api

**Pull request merged**